Controller fixes

garbear opened the pull request:
This PR manages the remaining bugs in the input system from #8807. I'll break out fixes in this PR to another PR as they're ready.

The remaining bugs are:
### Input
- [x] Joysticks not being detected on android startup (#10689)
- [x] Restart needed to enable/disable peripheral add-ons (#10690, #10734)
- [x] Missing lock in CAddonButtonMap (#10691)
- [x] Broken Shield controller if add-on isn't present (#10695)
- [x] Problems with accelerometers ([report and linux solution](http://forum.kodi.tv/showthread.php?tid=211138&pid=2207458#pid2207458)) (#10725, #10910)
- [x] PS4 controllers skip because button presses are sent when trigger is partially pressed (#10910)
- [x] Deadlock when updating peripheral add-ons (#10726)
- [x] Button map doesn't refresh when the user hits reset (#10735)
- [x] Possible crash when controllers are unplugged (#10740)
- [x] Incompatible with Unified Remote server ([report](http://forum.kodi.tv/showthread.php?tid=164725&pid=2119476#pid2119476), [support thread](http://forum.kodi.tv/showthread.php?tid=295961))
- [x] Constant joystick discovery on Windows ([report](http://forum.kodi.tv/showthread.php?tid=295548&pid=2449281#pid2449281), [peripheral.joystick#62](https://github.com/xbmc/peripheral.joystick/pull/62))
- [x] 100% cpu when scanning for joystick events after 6 days (#10921)
- [x] User should be able to open controller dialog with controllers unplugged (#10923)
- [x] Use 300ms rumble on weak motor for notifications (#10929)
- [ ] Incompatible with xpadder ([report](http://forum.kodi.tv/showthread.php?tid=173361&pid=2393609#pid2393609))
- [ ] All controllers vibrate whenever a new one is plugged in ([report](http://forum.kodi.tv/showthread.php?tid=173361&pid=2423447#pid2423447))
- [ ] Launching a game by holding A causes all hell to break loose ([report](http://forum.kodi.tv/showthread.php?tid=173361&pid=2413498#pid2413498))
- [ ] Missing a `mod="longpress"`  parameter for [joystick.xml](https://github.com/xbmc/xbmc/blob/master/system/keymaps/joystick.xml) keymap (should fix above bug)
- [ ] If Kodi crashes, the controller could vibrate without stopping ([report](http://forum.kodi.tv/showthread.php?tid=173361&pid=2430664#pid2430664))
- [ ] No navigation sounds for controllers ([report](http://forum.kodi.tv/showthread.php?tid=173361&pid=2442755#pid2442755))
- [ ] Mapping axis to Select sends one command per frame, locking the GUI ([sample buttonmap](https://gist.github.com/garbear/b1a8ce9eedc8bdcb462edab14a11c596))

### UX
- [x] Controller button should stay focused while mapping (#10761)
- [x] Open the controller dialog if a button press comes from an unmapped controller (#10882)
- [x] Volume is lowered when final analog stick is mapped ([report](http://forum.kodi.tv/showthread.php?tid=173361&pid=2430774#pid2430774), #10966)
- [x] Fix loss of control if A becomes unmapped (https://github.com/xbmc/peripheral.joystick/pull/65)
- [ ] Button list focus should stay centered:

<img width="1416" alt="screen shot 2016-10-09" src="https://cloud.githubusercontent.com/assets/531482/19223097/eab3cff0-8e1b-11e6-9898-376c7c5053ab.png">

In the image, you can see that the wizard focus travels down the page as the user maps buttons. The focus should stay in the center for as long as possible because this lets the user see which buttons are coming up.

garbear committed:
[joysticks] Only dispatch action if button was pressed this frame

This fixes a future problem when the "cheat code" to enable games
is merged. The cheat code is handled in the current frame, but in
the next frame, the button's Kodi action is executed.

This fixes the problem by only executing a Kodi action if the
button hasn't been held for more than 1 frame (hold time > 0).

garbear committed:
[input] Give priority to newly registered keyboard handlers

Currently, there is only one input handler: the controller dialog.
It intercepts keypresses durring button mapping to interrupt the
mapping wizard.

When RetroPlayer is merged, additional keyboard handles will be
introduced, like the one to monitor for the cheat code and the one
used for games requiring a keyboard.

Razzeee commented on xbmc/input/joysticks/KeymapHandler.cpp:
```diff
@@ -86,10 +86,14 @@ void CKeymapHandler::ProcessButtonPress(unsigned int keyId, unsigned int holdTim
 
-    if (SendDigitalAction(keyId))
+    // Only dispatch action if button was pressed this frame
+    if (holdTimeMs == 0)
```
You should be able to combine this if with the one below.

garbear committed:
[peripherals] Cleanup: Remove unused function

garbear commented:
addressed your comment @Razzeee. I also uploaded 3 more fixes from my retroplayer branch

HitcherUK commented:
Changing the lists to `fixedlist` and add 
```
<focusposition>4</focusposition>
<movement>4</movement>
```

should fix it for you

fetzerch committed:
[Docs] Add Doxygen documentation for input

garbear committed:
Refactor CKeymapHandler::GetInputType()

garbear committed:
Fix unneeded namespace identifier

garbear committed:
Fix crash on unplugging controller

garbear committed:
Only force-enable peripheral add-ons with button mapping support

garbear committed:
[joysticks] Cosmetic: fix comments

This fixes the bracket on the end of an interval and removes an
extraneous slash.

hudokkow commented on xbmc/input/joysticks/DeadzoneFilter.h:
```diff
@@ -45,7 +45,7 @@ namespace JOYSTICK
    *   - Negative in the interval [-1.0, -deadzone)
    *   - Zero in the interval [-deadzone, deadzone]
-   *   - Positive in the interval (deadzone, 1.0)
+   *   - Positive in the interval (deadzone, 1.0]
```
s/**(**deadzone, 1.0]/**[**deadzone, 1.0]

Montellese replied:
Nope, it's correct because it's a mathematical term specifying whether the value itself (in this case `1.0`) is inclusive or exlusive.

hudokkow replied:
Thx, learning something everyday. :wink:

garbear committed:
[joystick] Code improvements in DefaultJoystick.cpp/h

garbear committed:
[joysticks] Avoid sending 0.0 value actions every frame

garbear committed:
[joysticks] Fix no action when analog stick changes direction

garbear commented:
I'm pretty happy with the state that joystick are in for v17. jenkins build this please

hdkiller commented:
Hello!

Please add support for disabling rumble notifications.
